logstash-core/src/main/java/org/logstash/log/PipelineRoutingAppender.java

```diff
@@ -0,0 +1,153 @@
+ *            fileName="${sys:ls.logs}/pipeline_${ctx:pipeline.id}.log"
+ *            filePattern="${sys:ls.logs}/pipeline_${ctx:pipeline.id}.%i.log.gz">
+ *          <PatternLayout>
+ *            <pattern>[%d{ISO8601}][%-5p][%-25c] %m%n</pattern>
+ *          </PatternLayout>
+ *          <SizeBasedTriggeringPolicy size="100MB" />
+ *          <DefaultRolloverStrategy max="30" />
+ *        </RollingFile>
+ *  </PipelineRouting>
+ *  }
+ * </pre>
+ * */
+@Plugin(name = "PipelineRouting", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE, printObject = true, deferChildren = true)
+public class PipelineRoutingAppender extends AbstractAppender {
+
+    /**
+     * Builder for {@link PipelineRoutingAppender} instances
+     * */
+    public static class Builder<B extends PipelineRoutingAppender.Builder<B>> extends AbstractAppender.Builder<B>
+            implements org.apache.logging.log4j.core.util.Builder<PipelineRoutingAppender> {
+
+        @PluginNode
+        private Node appenderNode;
+
+        @Override
+        public PipelineRoutingAppender build() {
+            final String name = getName();
+            if (name == null) {
+                LOGGER.error("No name defined for this RoutingAppender");
+                return null;
+            }
+            return new PipelineRoutingAppender(name, appenderNode, getConfiguration());
+        }
+    }
+
+    /**
+     * Factory method to instantiate the appender
+     * */
+    @PluginBuilderFactory
+    public static <B extends PipelineRoutingAppender.Builder<B>> B newBuilder() {
+        return new PipelineRoutingAppender.Builder<B>().asBuilder();
+    }
+
+    private final Node appenderNode;
+    private final Configuration configuration;
+    private final ConcurrentMap<String, AppenderControl> createdAppenders = new ConcurrentHashMap<>();
+    private final Map<String, AppenderControl> createdAppendersUnmodifiableView =
+            Collections.unmodifiableMap(createdAppenders);
+
+    protected PipelineRoutingAppender(String name, Node appenderNode, Configuration configuration) {
+        super(name, null, null, false, new Property[0]);
+        this.appenderNode = appenderNode;
+        this.configuration = configuration;
+    }
+
+    /**
+     * Returns an unmodifiable view of the appenders created by this {@link PipelineRoutingAppender}.
+     */
+    public Map<String, AppenderControl> getAppenders() {
+        return createdAppendersUnmodifiableView;
+    }
+
+    /**
+     * Core method to apply the logic of routing.
+     * */
+    @Override
+    public void append(LogEvent event) {
+        AppenderControl appenderControl = getControl(event);
+
+        if (appenderControl != null) {
+            appenderControl.callAppender(event);
+        }
+    }
+
+    /**
+     * Create or retrieve the sub appender for the pipeline.id provided into the event
+     * */
+    private synchronized AppenderControl getControl(LogEvent event) {
```

Comment:
do we "really" need synchronization here, given that we're trying to use a concurrent map for the appenders?
maybe only sync the critical section where new appenders are created (with a double null check).
not sure it's worth the hustle (seemed like this code-path would be quite 'hot') - feel free to disregard.

Reply:
Nice catch, I took inspiration from original log4j's RoutingAppender, but your point is super good because on JDK15 biased locking was removed

Reply:
even though biased locking was removed in JDK15, we are going to continue to support JDK11 & friends for a while. Do we get the locking semantics we require pre-JDK15?

Reply:
The race condition is on create the appender iff is not present in the concurrent map, in respect to that the critical section covers this case. The map of appenders controls is a ConcurrentMap so it's safe to have a `get` outside of the critical section